### PR TITLE
chore: add integration tests using Playwright and Puppeteer

### LINKED ISSUE #556: Setup UI tests for the extension
We should be able to use Playwright, Puppeteer etc. and also hosted services to test UI flows.
I am opening this issue to collect some links for this. 

https://www.cypress.io/
https://github.com/microsoft/playwright
https://www.lambdatest.com/online-browser-testing
https://www.browserstack.com

### Cypress
+ https://www.cypress.io/
+ https://medium.com/@knoldus/how-to-load-chrome-extension-in-cypress-io-e313d3c68988

### Pupeteer

+ https://pptr.dev/
+ https://puppetry.app/
+ https://github.com/checkly/headless-recorder

### LambdaTest:
Support says they support extensions
"you can test through automation testing by uploading a file to browser> extenstion> unpacked or packed extension"
* https://www.lambdatest.com/support/api-doc/#/extensions/UploadExtensions

### PR DESCRIPTION
**Pull request by MiguelMedeiros**
Fix #556.

---

### New Dev Dependencies
This PR implements integration tests using:
- [Jest](https://jestjs.io/) 
- [Puppeteer](https://github.com/puppeteer/puppeteer)
- [Puppeteer Testing Library](https://testing-library.com/docs/)
- [Complete Randomer](https://www.npmjs.com/package/complete-randomer)

---

### Integrations tests
I've already added the following tests:
- [x] follow the welcome screen and connect using LNBits
- [x] follow the welcome screen and connect using BlueWallet
- [x] follow the welcome screen and create an Alby wallet

There are still more connections to the welcome screen, and other screens like spend and receive sats, connect and disconnect websites, lock and remove accounts. 

But we can do that in other PRs, the purpose of this PR is to start testing the extension so that the screens could have uniform paths and don't break due to any future change.

---

### How to test

Run this command to run all the tests.
```bash
yarn run test
```

---

## Preview

https://user-images.githubusercontent.com/5798170/155232637-eec1800d-f1a3-45aa-9363-c766a6bf7180.mp4

### CODE REVIEW

File: __tests__/integration/welcome/createWallets.test.tsx
Comment (jaonoctus) [inline __tests__/integration/welcome/createWallets.test.tsx:67]: nit: why not this (same applies for all other "then" calls)?

```suggestion
      const createNewWalletButton = await getByText($document, "Create a new wallet");
      
      createNewWalletButton.click();
```

Comment (MiguelMedeiros) [reply __tests__/integration/welcome/createWallets.test.tsx:67]: Nice feedback! 
I'll push a commit to normalize all the await calls! 

Thanks, bro! :heart:

**Review by jaonoctus — APPROVED** [review]
LGTM

**Review by escapedcat — COMMENTED** [review]
Really cool!  
Do you know if it's possible to record videos? Can be very helpful to create this when running in GH actions to see why an integration-test failed.

File: __tests__/integration/welcome/createWallets.test.tsx
Comment (MiguelMedeiros) [reply __tests__/integration/welcome/createWallets.test.tsx]: Sure! 
Fixed here: https://github.com/getAlby/lightning-browser-extension/pull/627/commits/b945f4512266754c81efa9d870344ee355e61629

File: __tests__/integration/welcome/createWallets.test.tsx
Comment (MiguelMedeiros) [reply __tests__/integration/welcome/createWallets.test.tsx]: I could not make it work with `getByRole`, do I need to add any property to the html tag?

**Comment by MiguelMedeiros** [comment]
> Really cool! Do you know if it's possible to record videos? Can be very helpful to create this when running in GH actions to see why an integration-test failed.

Good question. I do not know! :)
I could look into a tool for that (cypress has it by default).
Jest might have one too.

But I removed the `try/catch` code, so now you have the full messages when it fails.

File: __tests__/integration/welcome/createWallets.test.tsx
Comment (escapedcat) [reply __tests__/integration/welcome/createWallets.test.tsx]: Nope, should "just" work. If it fails it usually gives youa list of "roles" that were found. Maybe role "button" isn't correct here. But also jusy a nitpick. Don't spend too much time on it. We can optimize this later.

**Comment by escapedcat** [comment]
> Good question. I do not know! :) I could look into a tool for that (cypress has it by default). Jest might have one too.
I liked the cypress version because of that and the nice UI/UX for intergrationtests.  
What was your main reason to close that PR?  
If it's possible to have videos created with jest I guess that's also fine. Jest is faster in execution than cypress in my experience.

**Comment by escapedcat** [comment]
Switch to playwright and integration into CI looks good.  
@bumi @dylancom could you trigger the workflows here to see if this works?

File: __tests__/integration/welcome/createWallets.test.tsx
Comment (MiguelMedeiros) [reply __tests__/integration/welcome/createWallets.test.tsx]: OK, I will mark as resolved then! :)

**Comment by MiguelMedeiros** [comment]
> > Good question. I do not know! :) I could look into a tool for that (cypress has it by default). Jest might have one too.
> 
> I liked the cypress version because of that and the nice UI/UX for intergrationtests. What was your main reason to close that PR? If it's possible to have videos created with jest I guess that's also fine. Jest is faster in execution than cypress in my experience.

I think Jest is much faster then cypress and it is possible to integrate videos with Jest + Playwright! 
We can do that in another PR. :)

**Comment by MiguelMedeiros** [comment]
Hmm tests, failed. :/

I've removed a few `args  tags from the puppeteer launch. 
But Idk if it's going to fix the problem.

@escapedcat any ideas on how to fix this GH Action thing?

```bash
Error: Failed to launch the browser process!
[4433:4433:0225/070331.250413:ERROR:ozone_platform_x11.cc(234)] Missing X server or $DISPLAY
[4433:4433:0225/070331.250472:ERROR:env.cc(225)] The platform failed to initialize.  Exiting.


TROUBLESHOOTING: https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md


    at onClose (/home/runner/work/lightning-browser-extension/lightning-browser-extension/node_modules/puppeteer/src/node/BrowserRunner.ts:268:9)
    at ChildProcess.<anonymous> (/home/runner/work/lightning-browser-extension/lightning-browser-extension/node_modules/puppeteer/src/node/BrowserRunner.ts:255:61)
```

**Comment by escapedcat** [comment]
Does this help?:  
https://stackoverflow.com/a/61821201/1667461  
```js
await puppeteer.launch({
  headless: true,
  args: [
    "--disable-gpu",
    "--disable-dev-shm-usage",
    "--no-sandbox",
    "--disable-setuid-sandbox"
]
});
```

**Comment by MiguelMedeiros** [comment]
> Does this help?: https://stackoverflow.com/a/61821201/1667461
> 
> ```js
> await puppeteer.launch({
>   headless: true,
>   args: [
>     "--disable-gpu",
>     "--disable-dev-shm-usage",
>     "--no-sandbox",
>     "--disable-setuid-sandbox"
> ]
> });
> ```

The extension won't work with `headless: true`, take a look here:
https://github.com/puppeteer/puppeteer/issues/659#issuecomment-326754863

I've updated the args with the suggested tags.
I'm trying to setup a local GH actions at my forked repo.
Let me see if I could make it work.

**Comment by MiguelMedeiros** [comment]
Good! Tests are running, almost there! :) 

Error message:
```bash
FAIL tests/createWallets.spec.ts
  ● Test suite failed to run

    Playwright Test needs to be invoked via 'npx playwright test' and excluded from Jest test runs.
    Creating one directory for Playwright tests and one for Jest is the recommended way of doing it.
    See https://playwright.dev/docs/intro for more information about Playwright Test.

      104 |     const $document = await getDocument(page);
      105 |
    > 106 |     // go through welcome page
          |            ^
      107 |     const startedButton = await getByText($document, "Get Started");
      108 |     startedButton.click();
      109 |

      at throwIfRunningInsideJest (node_modules/@playwright/test/lib/testType.js:255:11)
      at TestTypeImpl._describe (node_modules/@playwright/test/lib/testType.js:88:5)
      at Function.describe (node_modules/@playwright/test/lib/transform.js:225:12)
      at Object.<anonymous> (tests/createWallets.spec.ts:106:12)
```

**Comment by escapedcat** [comment]
Awesome work @MiguelMedeiros !

Might be good to split linting and unit-test and intergartion-tests into individual checks in the future to indicate quicker where a certain error is coming from.

File: .github/workflows/playwright.yml
Comment (escapedcat) [inline .github/workflows/playwright.yml:35]: Does `yarn` work here or does it have to be `npx` because some things aren't installed yet?  
Background is that everything was aligned to use `yarn` to be consistent.

Comment (MiguelMedeiros) [reply .github/workflows/playwright.yml:35]: Yeah, I couldn't make it work using `yarn` to start playwright.
For some reason `yarn` was calling jest instead of playwright, and there was a warning to use `npx` instead.

The only problem right now is that we need to use `headless: false` when we launch the puppeteer browser to test the extension. And GH Actions is not running the e2e tests properly because of that. 
That is why I used: `mujo-code/puppeteer-headful@v2`.

I wonder what error msg GH Actions will show to us. Can we run the tests again?

Comment (bumi) [reply .github/workflows/playwright.yml:35]: @MiguelMedeiros sorry for the limitation. actually only one approval should be needed per contributor. I've also added `mujo-code/puppeteer-headful@v2` to the allowed actions. 
I had this limited initially when I've setup some repository secrets.

**Comment by bumi** [comment]
Note: we need to configure the repository to use a test endpoint to create wallets. Right now we still create production wallets on each run. (maybe good for the user stats 🤔 )

**Comment by escapedcat** [comment]
> Note: we need to configure the repository to use a test endpoint to create wallets. 

@MiguelMedeiros have a look at these:  https://github.com/getAlby/lightning-browser-extension/wiki/Test-setup

**Comment by bumi** [comment]
@kiwiidb let's setup getalby.com for the regtest nodes and then we can use that one in the tests. 
I am hesitant to merge this before as the tests currently run against production.

**Comment by escapedcat** [comment]
@bumi what's missing here? Would be great to get this in to improve testing

**Comment by bumi** [comment]
yep. only missing that tests run against the test environment. (otherwise always new production accounts are created)

**Comment by kiwiidb** [comment]
Currently we are blocked by the regtest Ruby app throwing errors with regards to the LNDhub connection.
I've looked into this multiple times, but I don't really know how to start debugging a Ruby app...
The app url is https://app.regtest.getalby.com, and the lndhub url is https://lndhub.regtest.getalby.com. The test suite could already be pointed towards these endpoints instead of to the mainnet URL.

**Comment by bumi** [comment]
what is the best to set the URL (https://github.com/getAlby/lightning-browser-extension/blob/master/src/app/screens/Onboard/NewWallet/index.tsx#L11) ? should we default to use the regtest setup or set an environment variable in the tests?

**Comment by kiwiidb** [comment]
https://app.regtest.getalby.com should be fully operational and ready to be used in the test suite now.

**Comment by escapedcat** [comment]
@MiguelMedeiros thanks for this! Can't wait to use this and add more tests.

I cloned and checked out the branch to give this a try. I used node v14. 

Got:  
```sh
Error: listen EADDRINUSE: address already in use :::9323
```

I'm not aware of any other running process blocking this.  
Any idea what I missed?

Other questions/feedback:
- What needs to be done before running `yarn playwright test` to make this work? And can we add needed requirements to the readme?
  - Install chromium?
  - Run `yarn run dev:chrome`?
  - Install the Load unpacked extension into Chromium?
- `yarn test` (as mentioned above) says:
   > `Playwright Test needs to be invoked via 'npx playwright test' and excluded from Jest test runs.`

File: README.md
Comment (escapedcat) [inline README.md:110]: Maybe describe previous necessary steps which might be needed.  

Does it make sense to split this up into "Unit tests" and "Integration/e2e tests"?

File: README.md
Comment (escapedcat) [inline README.md:113]: Can also be `npx` then I think

File: jest.config.ts
Comment (escapedcat) [inline jest.config.ts:13]: Is this (still) needed?  
Or this needed for CI only? If yes, can this be smaller? Otherwise it might take forever till the pipeline eventually fails.

File: playwright.config.ts
Comment (escapedcat) [inline playwright.config.ts:14]: Is tests confusing regarding the type of tests? Unit, intergration, e2e? Maybe a sub-folder would work?

File: playwright.config.ts
Comment (escapedcat) [inline playwright.config.ts:16]: What's the default timeout? Do we need 30secs? Can it be smaller?

File: playwright.config.ts
Comment (escapedcat) [inline playwright.config.ts:35]: `0` is already the default, right? Can be commented out again?

**Comment by bumi** [comment]
what are we missing here? let's get this in. 
@kiwiidb can you help with making sure tests run against our test environment?

**Review by kiwiidb — COMMENTED** [review]
We can easily test LND and even Eclair wallet creation as well with the regtest setup.

File: tests/createWallets.spec.ts
Comment (kiwiidb) [inline tests/createWallets.spec.ts:42]:
```suggestion
    process.env['WALLET_CREATE_URL'] = 'https://app.regtest.getalby.com/api/users';
    const user = USER.SINGLE();
```

Comment (bumi) [reply tests/createWallets.spec.ts:42]: setting the environment variable should probably not happen in the test case because this has global side effects, and is not only changing the env variable for the specific test case. we should set this globally for test runs.

Comment (kiwiidb) [reply tests/createWallets.spec.ts:42]: So we better set this in the `loadExtension` function?

Comment (kiwiidb) [reply tests/createWallets.spec.ts:42]: Setting it in the Github pipeline is not a great idea because someone that runs the tests locally should also connect to the regtest environment when creating wallets. Another option is to have the default value _in the code_ point to the regtest URL, and override this somehow in the main process with the production host.

Comment (bumi) [reply tests/createWallets.spec.ts:42]: +1 for regest defaults. we overwrite env variables currently already for production builds.
can you update it?

Comment (bumi) [reply tests/createWallets.spec.ts:42]: I don't know the `loadExtension` function. but if you set a global variable in a test it has an effect on all the other tests. - which is unintended. 
tests run (hopefully) in random order thus this will also only sometimes influence a test.

Comment (kiwiidb) [reply tests/createWallets.spec.ts:42]: How can I add a commit to this PR since the base is coming from MiguelMedeiros' fork?

Comment (kiwiidb) [reply tests/createWallets.spec.ts:42]: Or should I check this out in a new branch and make a new PR?

Comment (bumi) [reply tests/createWallets.spec.ts:42]: you should be able to add commits to this PR (this branch of his fork) GH allows that for maintainers.

Comment (kiwiidb) [reply tests/createWallets.spec.ts:42]: I'm not allowed to do that.
```
To github.com:MiguelMedeiros/lightning-browser-extension.git
 ! [remote rejected]   MiguelMedeiros/test/integration-tests -> 
MiguelMedeiros/test/integration-tests (permission denied)
error: failed to push some refs to 'github.com:MiguelMedeiros/lightning-browser-extension.git'
```

Comment (bumi) [reply tests/createWallets.spec.ts:42]: interesting. actually that's default normally. I also use the `gh` CLI to checkout and push.

Comment (escapedcat) [reply tests/createWallets.spec.ts:42]: Just pushed a docs change to test this. Works for me. I'm not using the Github CLI. Just plain git.  
I cloned his repo, checked out the branch, created a commit and pushed. 🤷‍♂️

**Comment by escapedcat** [comment]
Rebased with master to resolve conflicts. Playwright failing now, yay...  
Will try to find some time to understand this, make it run locally and wrap this up. Would be really great to have this.

**Comment by kiwiidb** [comment]
> Just pushed a docs change to test this. Works for me. I'm not using the Github CLI. Just plain git. I cloned his repo, checked out the branch, created a commit and pushed. 🤷‍♂️

This worked for me as well. I've changed the default wallet create URL and was able to create a new wallet in the extension using `app.regtest.getalby.com`. Just one question, where is it changed again to point to the production endpoint?

File: README.md
Comment (escapedcat) [inline README.md:35]: Ugh, that's my fault, will fix

**Comment by escapedcat** [comment]
>  Just one question, where is it changed again to point to the production endpoint?

This? https://github.com/getAlby/lightning-browser-extension/pull/627/files#diff-070b2745eed6a27072e7e29f4fd76b750052f35bb9a3764a80b5d0bbf6b78586L11-R11

**Comment by kiwiidb** [comment]
> > Just one question, where is it changed again to point to the production endpoint?
> 
> @kiwiidb This? https://github.com/getAlby/lightning-browser-extension/pull/627/files#diff-070b2745eed6a27072e7e29f4fd76b750052f35bb9a3764a80b5d0bbf6b78586L11-R11

That's what I just changed (bc [@bumi agreed](https://github.com/getAlby/lightning-browser-extension/pull/627#discussion_r854217479)) , yes. So what I'm asking is where `WALLET_CREATE_URL` is set to the appropriate value for prod (ie. what the default was before I changed it).

**Comment by bumi** [comment]
when running `yarn package` the required environment variables must be set (e.g. `WALLET_CREATE_URL`) 
`WALLET_CREATE_URL=ff NODE_ENV=production FAUCET_URL=... yarn package`

**Comment by bumi** [comment]
(the gh CLI does nothing differently, just checkout the fork and push there, but it gives you a nice `gh pr checkout 627` so you only need to know the PR number)

File: tests/e2e/createWallets.spec.ts
Comment (escapedcat) [inline tests/e2e/createWallets.spec.ts:14]: Had to modify the ID to make it work.
@bumi do you know when this changes? Depending on that we need to think about how to handle this a somewhat automatic way in the future.

File: tests/e2e/createWallets.spec.ts
Comment (escapedcat) [inline tests/e2e/createWallets.spec.ts:41]: For now I reduced it to this test only so we can test it faster.

**Review by escapedcat — COMMENTED** [review]
Made it run locally after some changes.  
The tests fails because UI changed I believe. Will have a look at that next.

File: README.md
Comment (escapedcat) [inline README.md:132]: We could run all tests here and split it internally to have i.e.
- `yarn test:jest`
- `yarn test:playwright`

**Review by escapedcat — COMMENTED** [review]
Tests is updated and passes.
Give it a try locally. We might need to update the README to make everything as simple as possible.

File: .github/workflows/playwright.yml
Comment (escapedcat) [inline .github/workflows/playwright.yml:24]: @bumi also getting this:
```sh
orales/actions-yarn@v2.3.0 is not allowed to be used in getAlby/lightning-browser-extension. Actions in this workflow must be: within a repository owned by getAlby, created by GitHub, or matching the following: peter-evans/create-or-update-comment@v1, peter-evans/find-comment@v1, release-drafter/release-drafter@v5, zdurham/s3-upload-github-action@master, mujo-code/puppeteer-headful@v2.
```

File: tests/e2e/createWallets.spec.ts
Comment (bumi) [reply tests/e2e/createWallets.spec.ts:14]: I think on chrome development this changes every new install? or at least is different on every computer.
Only the production one is fixed afaik.

File: README.md
Comment (bumi) [reply README.md:132]: +1

File: .github/workflows/playwright.yml
Comment (bumi) [reply .github/workflows/playwright.yml:24]: added `borales/actions-yarn` to the list (and allow verified marketplace actions now) 
I've limited it for additional security reasons.

Comment (escapedcat) [reply .github/workflows/playwright.yml:24]: Hm, it's still not running:  
https://github.com/getAlby/lightning-browser-extension/actions/runs/2256837754  
Any idea?

Comment (escapedcat) [reply .github/workflows/playwright.yml:24]: Let me check if I can add missing ones

Comment (escapedcat) [reply .github/workflows/playwright.yml:24]: It's runing.  
Needed to be `borales/actions-yarn@v2.3.0` instead of `borales/actions-yarn` 🤦‍♂️

**Comment by escapedcat** [comment]
🎉  progress, we got a failing test!  
Need to debug and fix that now :P

**Comment by escapedcat** [comment]
As long as it's not running in the CI let's keep this as draft.  
Will have a look at #624. Might be easier to get that running in the CI.

**Comment by escapedcat** [comment]
After a little bit a back and forth this turned green. Can't believe it.  
Need to clean and update things a bit, but have a look.

**Comment by bumi** [comment]
whooot! 💚 💚 💚

File: package.json
Comment (escapedcat) [inline package.json:31]: Not sure how helpful these are. A specific path can be added to any script.

File: tests/e2e/createWallets.spec.ts
Comment (escapedcat) [inline tests/e2e/createWallets.spec.ts:117]: This is not working with regtest.

File: .github/workflows/codequality.yml
Comment (escapedcat) [inline .github/workflows/codequality.yml:1]: Kinda want to split this fiel up into 3 single ones to make it easier to see what exactly fails on a PR. What do you think?

File: src/app/screens/connectors/NewWallet/index.tsx
Comment (bumi) [reply src/app/screens/connectors/NewWallet/index.tsx:11]: Yes, I think we should keep regtest there as default. One uses dev/CI more often than a production build. 

If we set it to regtest then we only have to adjust the production builds to use the `WALLET_CREATE_URL` for production?

File: tests/e2e/createWallets.spec.ts
Comment (bumi) [inline tests/e2e/createWallets.spec.ts:57]: maybe we call this `welcomePage` to be more descriptive about which page this is? 

or we allow passing in an argument to `loadExtension` of the page that should be loaded?

**Review by bumi — COMMENTED** [review]
👍 thanks for getting this working!

File: .github/workflows/codequality.yml
Comment (escapedcat) [reply .github/workflows/codequality.yml:1]: Can be done via #912

File: src/app/screens/connectors/NewWallet/index.tsx
Comment (escapedcat) [reply src/app/screens/connectors/NewWallet/index.tsx:11]: I think we should only use the PROD-URL when we create a release.
For tests in the Github workflow we should use a production build with the REGTEST-URL. Same for production build you want to test locally.

File: tests/e2e/createWallets.spec.ts
Comment (escapedcat) [reply tests/e2e/createWallets.spec.ts:115]: Let's have a look together. I recently had issues with this and wasn't sure what the problem is.

Comment (escapedcat) [reply tests/e2e/createWallets.spec.ts:63]: It's this:
image
@kiwiidb is this still the DNS issue?

Comment (kiwiidb) [reply tests/e2e/createWallets.spec.ts:63]: This host should only be called from inside the cluster. I might have misconfigured this. Where is this called?

Comment (escapedcat) [reply tests/e2e/createWallets.spec.ts:66]: @kiwiidb opened a ticket at DO for the DNS issue.  
Reduced the test further for now. See above comment:
```js
    // @TODO: we have an DNS issue
    // a ticket was opened at DigitalOcean
```

Comment (bumi) [reply tests/e2e/createWallets.spec.ts:66]: I think this is wrong. we use an invalid URL here. This looks to me as a cluster internal URL. 
The internal DNS issue should have any effect here, or why does it?

File: package.json
Comment (bumi) [inline package.json:32]:
```suggestion
    "test": "yarn run test:unit && yarn run test:e2e",
```

Comment (escapedcat) [reply package.json:32]: It's actually not needed when using `yarn`. From my experience omitting this is the default when using `yarn`.

File: tests/e2e/createWallets.spec.ts
Comment (escapedcat) [reply tests/e2e/createWallets.spec.ts:66]: We talked. Closing this.

File: package.json
Comment (escapedcat) [reply package.json:32]: Ah, sorry..... correcting!

**Review by bumi — APPROVED** [review]

**Comment by escapedcat** [comment]
Thanks @MiguelMedeiros for starting this! ❤️

**Pull request merged**
